narwhal/primary/src/primary.rs

```diff
@@ -712,67 +711,26 @@ impl PrimaryReceiverHandler {
 
-        // Clone the round updates channel so we can get update notifications specific to
-        // this RPC handler.
-        let mut rx_narwhal_round_updates = self.rx_narwhal_round_updates.clone();
-        // Maximum header age is chosen to strike a balance between allowing for slightly older
-        // certificates to still have a chance to be included in the DAG while not wasting
-        // resources on very old vote requests. This value affects performance but not correctness
-        // of the algorithm.
-        const HEADER_AGE_LIMIT: Round = 3;
-
         // If requester has provided us with parent certificates, process them all
         // before proceeding.
         self.metrics
             .certificates_in_votes
             .inc_by(request.body().parents.len() as u64);
-        let mut wait_notifications: FuturesUnordered<_> = request
-            .body()
-            .parents
-            .clone()
-            .into_iter()
-            .map(|cert| self.synchronizer.wait_to_accept_certificate(cert))
-            .collect();
-        loop {
-            tokio::select! {
-                result = wait_notifications.next() => {
-                    match result {
-                        Some(result) => {
-                            result?;
-                            continue;
-                        },
-                        // Waits are done. Missing parents have been accepted.
-                        None => break,
-                    }
-                },
-                result = rx_narwhal_round_updates.changed() => {
-                    if result.is_err() {
-                        return Err(DagError::NetworkError("Node shutting down".to_owned()));
-                    }
-                    let narwhal_round = *rx_narwhal_round_updates.borrow();
-                    ensure!(
-                        narwhal_round.saturating_sub(HEADER_AGE_LIMIT) <= header.round(),
-                        DagError::TooOld(header.digest().into(), header.round(), narwhal_round)
-                    )
-                },
-            }
-        }
 
-        // Ensure we have the parents. If any are missing, the requester should provide them on retry.
-        // This check is necessary for correctness, because it is possible that the list of missing
-        // parents in the request is incomplete, or wait_notifications get notified on shut down
-        // without actually having the parents available.
-        let (parents, missing) = self.synchronizer.get_parents(header)?;
-        if !missing.is_empty() {
-            return Ok(RequestVoteResponse {
-                vote: None,
-                missing,
-            });
-        }
+        // Ensure the header has all parents accepted. If some are missing, waits until they become
+        // available or this request times out / is cancelled.
+        // This check is necessary for correctness.
+        let parents = self.synchronizer.wait_for_parents(header).await?;
```

Comment:
Not for this PR , but assuming we change the Certificate structure, I believe if we piggy back the authority id to the header parents we can leverage `certificate_fetcher` to kick off the sync mechanism for missing parents - assuming those are not already in flight. My understanding is that It seems that we do have a suspicion that we are dealing with some sort of contention (it's surprising if currently `synchronizer` is dealing with such issues as most of the operations are in-memory?) . Doing so will provide stronger guarantees that we'll eventually fetch any missing certificates instead of only relying to the consistent broadcast of last certificates mechanism only. I don't know if that will still contribute to some amplification , but it might worth the try as an experiment

Reply:
I think we eventually do want targeted certificate fetch. But I'm not sure if we want certificate sync to be triggered solely by header proposals. It could be a future direction.